asserts,image: add support for models with bases

mvo5 opened the pull request:
This adds support for model assertions that contain an optional `base: snapName` field. When those are found the image will be using this base instead of the "core" snap and try to boot from it.

zyga commented on asserts/device_asserts_test.go:
```diff
@@ -121,6 +123,21 @@ func (mods *modelSuite) TestDecodeStoreIsOptional(c *C) {
+	a, err = asserts.Decode([]byte(encoded))
+	c.Assert(err, IsNil)
+	model = a.(*asserts.Model)
+	c.Check(model.Base(), Equals, "")
```
Should this return "core" for simplicity?

pedronis replied:
all of asserts so far doesn't know about "core" the snap name,  I wouldn't add it here even now

zyga commented on image/image.go:
```diff
@@ -311,8 +312,12 @@ func bootstrapToRootDir(tsto *ToolingStore, model *asserts.Model, opts *Options,
-	// core,kernel,gadget first
-	snaps = append(snaps, local.PreferLocal(defaultCore))
+	// core/base,kernel,gadget first
+	if model.Base() != "" {
```
The part to tweak if you choose to handle this inside the assertion code.

zyga commented on image/image.go:
```diff
@@ -370,8 +375,8 @@ func bootstrapToRootDir(tsto *ToolingStore, model *asserts.Model, opts *Options,
-		// kernel/os are required for booting
-		if typ == snap.TypeKernel || typ == snap.TypeOS {
+		// kernel/os/model.base are required for booting
+		if typ == snap.TypeKernel || typ == snap.TypeOS || snapName == model.Base() {
```
What if model.Base returns "" ?

mvo5 replied:
That should be ok, as long as snapName is not empty. I will add code to ensure this is never the case.

pedronis replied:
you could capture baseName set to defaultCore or model.Base() != "" earlier instead

zyga commented on image/image.go:
```diff
@@ -464,12 +477,18 @@ func setBootvars(downloadedSnapsInfo map[string]*snap.Info) error {
+		info := downloadedSnapsInfoForBootConfig[fn]
 		switch info.Type {
-		case snap.TypeOS:
+		// FIXME: workaround a store bug thatsends
```
Typo `that sends`

bboozzoo replied:
`s/thatsends/that sends/`

zyga commented on image/image.go:
```diff
@@ -464,12 +477,18 @@ func setBootvars(downloadedSnapsInfo map[string]*snap.Info) error {
-		case snap.TypeOS:
+		// FIXME: workaround a store bug thatsends
+		// "type: base" as "application" :/
+		// Once the store is fixed the next two lines can be removed
```
This will be easier to remove if you include the URL of the bug.

mvo5 replied:
Thanks, fixed by removing the code (store got fixed since)

zyga commented on tests/main/prepare-image-grub-core18/task.yaml:
```diff
@@ -0,0 +1,81 @@
+    GADGET: /tmp/root/gadget
+    STORE_DIR: $(pwd)/fake-store-blobdir
+    STORE_ADDR: localhost:11028
+    UBUNTU_IMAGE_SKIP_COPY_UNVERIFIED_SNAPS: 1
```
Hmm?

bboozzoo replied:
That's probably some magical switch to `ubuntu-image`.

mvo5 replied:
We don't need this for this particular test, I removed it.

zyga commented on tests/main/prepare-image-grub-core18/task.yaml:
```diff
@@ -0,0 +1,81 @@
+    if [ "$REMOTE_STORE" = production ]; then
+        core18_snap_id="CSO04Jhav2yK0uz97cr0ipQRyqg0qQL6"
+    else
+        core18_snap_id="FIXME"
```
Heh, nice

bboozzoo commented on tests/main/prepare-image-grub-core18/task.yaml:
```diff
@@ -0,0 +1,81 @@
+        exit
+    fi
+
+    . $TESTSLIB/store.sh
```
`#shellcheck tests/lib/store.sh`

mvo5 replied:
Thanks, made it shellcheck clean now.

bboozzoo commented on tests/main/prepare-image-grub-core18/task.yaml:
```diff
@@ -0,0 +1,81 @@
+        exit
+    fi
+
+    . $TESTSLIB/store.sh
```
same as above

bboozzoo commented on tests/main/prepare-image-grub-core18/task.yaml:
```diff
@@ -0,0 +1,81 @@
+
+    . $TESTSLIB/store.sh
+    teardown_fake_store $STORE_DIR
+    rm -rf $ROOT
```
shellcheck will probably complain here, feel free to disable it

bboozzoo commented on tests/main/prepare-image-grub-core18/task.yaml:
```diff
@@ -0,0 +1,81 @@
+    su -c "SNAPPY_USE_STAGING_STORE=$SNAPPY_USE_STAGING_STORE snap prepare-image --channel edge --extra-snaps snapweb $TESTSLIB/assertions/developer1-pc-w-base.model $ROOT" test
+
+    echo Verifying the result
+    ls -lR $IMAGE
```
`"$IMAGE"`

bboozzoo commented on tests/main/prepare-image-grub-core18/task.yaml:
```diff
@@ -0,0 +1,81 @@
+    echo Verifying the result
+    ls -lR $IMAGE
+    for f in pc pc-kernel core18 snapd snapweb; do
+        ls $IMAGE/var/lib/snapd/seed/snaps/${f}*.snap
```
same as above

bboozzoo commented on tests/main/prepare-image-grub-core18/task.yaml:
```diff
@@ -0,0 +1,81 @@
+
+    echo Expose the needed assertions through the fakestore
+    cp $TESTSLIB/assertions/developer1.account $STORE_DIR/asserts
+    cp $TESTSLIB/assertions/developer1.account-key $STORE_DIR/asserts
```
I think shellcheck will complain about quoting those

pedronis commented on asserts/device_asserts.go:
```diff
@@ -81,6 +81,11 @@ func (mod *Model) Kernel() string {
 }
 
+// Base returns the base snap the model uses.
+func (mod *Model) Base() string {
```
I think we should prevent to set base if the model is classic.  If understand the snapd snap will not have a base, so a classic system doesn't have a base, it will just fetch the bases needed for the snaps.

pedronis commented on boot/kernel_os.go:
```diff
@@ -147,6 +148,16 @@ func SetNextBoot(s *snap.Info) error {
+		// Nothing to do, this is just a base snap but not
+		// the base name we used for booting.
+		return nil
+	}
```
mmh, this check in this form feels a bit of a hack, wondering if we can do better

pedronis replied:
notice that devicestate already sets various helpers for use in snapstate,   it could also set some helper to know what is the system model base

mvo5 replied:
I agree, using the devicestate.Model() is best here. I can tweak this. One reason it is done the way it is right now is that we don't have a core18 model so in order to test refreshes from core18 -> core18 this is the most practical way right now. But I will look into make this better.

pedronis commented on boot/kernel_os.go:
```diff
@@ -155,7 +166,7 @@ func SetNextBoot(s *snap.Info) error {
 // KernelOrOsRebootRequired returns whether a reboot is required to swith to the given OS or kernel snap.
 func KernelOrOsRebootRequired(s *snap.Info) bool {
-	if s.Type != snap.TypeKernel && s.Type != snap.TypeOS {
+	if s.Type != snap.TypeKernel && s.Type != snap.TypeOS && s.Type != snap.TypeBase {
```
this is correct for unrelated bases because of the check on nextBoot, feels still a bit fragile, either way the unrelated  base case merits new/own tests I think

pedronis commented on boot/kernel_os_test.go:
```diff
@@ -181,6 +181,50 @@ func (s *kernelOSSuite) TestSetNextBootForCore(c *C) {
+	c.Assert(err, IsNil)
+
+	c.Assert(s.bootloader.BootVars, DeepEquals, map[string]string{
+		// Note that snap_try_core and snap_mode are unset
```
they could be set though,  a test for that case too would be good

pedronis replied:
was this case addressed? I don't see a new test

pedronis commented on image/image.go:
```diff
@@ -402,6 +411,10 @@ func bootstrapToRootDir(tsto *ToolingStore, model *asserts.Model, opts *Options,
 	}
+	// XXX: make this a warning only?
+	if model.Base() != "" && !seen["snapd"] {
+		return fmt.Errorf(`Base %q is used but no "snapd" snap is installed`, model.Base())
```
forcing people to add snapd themselves seems strange, see same argument on the forum

pedronis commented on image/image.go:
```diff
@@ -323,9 +328,13 @@ func bootstrapToRootDir(tsto *ToolingStore, model *asserts.Model, opts *Options,
+	downloadedSnapsInfoForBootConfig := map[string]*snap.Info{}
 	var seedYaml snap.Seed
 	for _, snapName := range snaps {
+		if snapName == "" {
```
if we want this check, but see comment below, it should happen earlier than here I think

mvo5 replied:
I removed this from this PR again and moved the capturing of baseName some lines earlier.

pedronis commented on image/image.go:
```diff
@@ -384,7 +393,7 @@ func bootstrapToRootDir(tsto *ToolingStore, model *asserts.Model, opts *Options,
 			}
 			// store the snap.Info for kernel/os so
```
/base ?

mvo5 commented:
@pedronis Thanks for the review. I refactored the code now so that SetNextBoot is only called for the kernel/base/core as defined in the model assertion. I hope this addresses the concerns about fragile logic.

pedronis commented on overlord/snapstate/backend/link.go:
```diff
@@ -67,8 +68,17 @@ func (b Backend) LinkSnap(info *snap.Info) error {
 	// XXX/TODO: this needs to be a task with proper undo and tests!
-	if err := boot.SetNextBoot(info); err != nil {
-		return err
+	if model != nil {
```
given we moved some of the checking here, this should probably be guarded by !classic as well

pedronis commented on boot/kernel_os.go:
```diff
@@ -101,14 +100,16 @@ func ExtractKernelAssets(s *snap.Info, snapf snap.Container) error {
+// the right bootable base (from the model assertion).
 func SetNextBoot(s *snap.Info) error {
 	if release.OnClassic {
 		return nil
```
should this also be an error now for consistency?

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -579,7 +585,8 @@ func (m *SnapManager) undoUnlinkCurrentSnap(t *state.Task, _ *tomb.Tomb) error {
 
 	snapst.Active = true
-	err = m.backend.LinkSnap(oldInfo)
+	model, _ := Model(st)
```
we shouldn't ignore the error when is not ErrNoState I think

pedronis commented on image/image.go:
```diff
@@ -310,11 +311,20 @@ func bootstrapToRootDir(tsto *ToolingStore, model *asserts.Model, opts *Options,
 	snaps = append(snaps, local.PreferLocal(model.Gadget()))
+	// always add an implicit snapd when a base is used
+	if model.Base() != "" {
+		snaps = append(snaps, "snapd")
```
it needs to be the first one I think or there will be problems with hooks of gadget etc
